execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -1067,25 +1070,39 @@ where
             mut path,
         }) = self.visited.pop()
         {
-            let mut maybe_next_trie: Option<Trie<K, V>> = None;
+            let mut maybe_next_trie: Option<LazyTrieLeaf<K, V>> = None;
 
             match trie {
-                Trie::Leaf { key, .. } => {
-                    let key_bytes = match key.to_bytes() {
-                        Ok(bytes) => bytes,
-                        Err(e) => {
-                            self.state = KeysIteratorState::Failed;
-                            return Some(Err(e.into()));
-                        }
-                    };
-                    debug_assert!(key_bytes.starts_with(&path));
+                LazyTrieLeaf::Left(leaf_bytes) => {
+                    if leaf_bytes.is_empty() {
+                        self.state = KeysIteratorState::Failed;
+                        return Some(Err(bytesrepr::Error::Formatting.into()));
+                    }
+
+                    let key_bytes = &leaf_bytes[1..]; // Skip `Trie::Leaf` tag
```

Comment:
nit: maybe it would be good to add a comment to `ToBytes`/`FromBytes` in `Trie` to point out that there are places where we do piecewise deserialization of it. It's good to know if we ever change that implementation that we also need to re-factor this code for it to work correctly.

Reply:
Added comment